### PR TITLE
[ML] Improve boosted tree training initialisation

### PR DESCRIPTION
**Pull request by tveasey**
This makes some changes to initialisation:
1. Measures the gain and sum curvature in the tree directly to estimate upper bounds for good values for `gamma` and `lambda`, respectively,
2. Search a large range of values from these initial overestimates looking for a turning point in the test error as the model transitions from underfit to overfit.

The hyperaparameter search is centred on the values at this transition. I've reduced the number of hyperparameter optimisation rounds as a result of the improved initialisation.

This also does a better job of monitoring progress. We explicitly account for the cost of initialisation and update progress after each forest is trained rather than per round of the hyperparameter optimisation. Since it is useful to share progress monitoring between the boosted tree factory and the implementation I've migrated to storing the loop progress monitor on the implementation and persisting and restoring. This incidentally fixes a bug in progress monitoring on resume.

Finally, I've refactored the regularisation parameters to better encapsulate. This anticipates depth based regularisation.

### CODE REVIEW

File: lib/core/CLoopProgress.cc [hunk 1]
```diff
@@ -30,5 +50,40 @@ void CLoopProgress::increment(std::size_t i) {
         m_LastProgress += stride;
     }
 }
+
+void CLoopProgress::resumeRestored() {
+    this->increment(0);
```
Comment (valeriy42) [inline lib/core/CLoopProgress.cc]: Does this do anything?

File: include/maths/CBoostedTreeImpl.h [hunk 2]
```diff
@@ -120,14 +120,73 @@ class MATHS_EXPORT CBoostedTreeImpl final {
     using TNodeVec = std::vector<CNode>;
     using TNodeVecVec = std::vector<TNodeVec>;
 
+    //! \brief Holds the parameters associated with the different types of regulariser
+    //! terms available.
+    template<typename T>
+    class CRegularization final {
+    public:
+        //! Set the multiplier of the tree size regularizer.
```
Comment (valeriy42) [inline include/maths/CBoostedTreeImpl.h]: I guess it is "multiplier of the tree size penalty". Same also below.

File: include/maths/CBoostedTreeImpl.h [hunk 3]
```diff
@@ -662,9 +676,8 @@ class MATHS_EXPORT CBoostedTreeImpl final {
 
     //! Compute the sum loss for the predictions from \p frame and the leaf
     //! count and squared weight sum from \p forest.
-    TDoubleDoubleDoubleTr regularisedLoss(const core::CDataFrame& frame,
-                                          const core::CPackedBitVector& trainingRowMask,
-                                          const TNodeVecVec& forest) const;
+    TDoubleDoublePr gainAndCurvatureAtPercentile(double percentile,
```
Comment (valeriy42) [inline include/maths/CBoostedTreeImpl.h]: Function documentation has to be updated.

File: lib/maths/CBoostedTreeFactory.cc [hunk 4]
```diff
@@ -334,6 +482,8 @@ CBoostedTreeFactory::constructFromString(std::istream& jsonStringStream,
         if (treePtr->acceptRestoreTraverser(traverser) == false || traverser.haveBadState()) {
             throw std::runtime_error{"failed to restore boosted tree"};
         }
+        treePtr->m_Impl->m_TrainingProgress.attach(recordProgress);
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeFactory.cc]: And indeed that was what was causing the test failures!

File: lib/core/CLoopProgress.cc (hunk 1 above)
Comment (tveasey) [reply lib/core/CLoopProgress.cc]: It refreshes the `m_LastProgress` which *is not* persisted. This triggers us also to output all the progress we made up to the point that persist was called. I think this probably warrants some comments, also in persist.

File: lib/maths/CBoostedTreeFactory.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeFactory.cc]: (I'm mulling over having `constructFromString` and return the factory as we do from construct from parameters. Aesthetically, I don't like the asymmetry. I'm just seeing how it'll work out.)

File: include/maths/CBoostedTreeImpl.h (hunk 2 above)
Comment (tveasey) [reply include/maths/CBoostedTreeImpl.h]: +1

File: lib/maths/CBoostedTreeFactory.cc [hunk 5]
```diff
@@ -315,6 +280,189 @@ void CBoostedTreeFactory::initializeHyperparameters(core::CDataFrame& frame) con
...
+    double curvature;
+    std::tie(gain, curvature) = m_TreeImpl->gainAndCurvatureAtPercentile(75.0, forest);
+
+    LOG_TRACE(<< "gain = " << gain << ", curvature = " << curvature);
+
+    return {gain, curvature};
+}
+
+CBoostedTreeFactory::TOptionalVector
+CBoostedTreeFactory::candidateRegularizerSearchInterval(core::CDataFrame& frame,
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeFactory.cc]: Exactly and thanks! The problem I had was I need a different setter on the regularisation object. That said, I think I can make this much clearer along the lines you suggest and also expose the hidden dependency on `m_TreeImpl`.

File: lib/maths/CBoostedTreeFactory.cc [hunk 6]
```diff
@@ -78,13 +88,13 @@ void CBoostedTreeFactory::initializeHyperparameterOptimisation() const {
     // less than p_1, this translates to using log parameter values.
 
     CBayesianOptimisation::TDoubleDoublePrVec boundingBox;
-    if (m_TreeImpl->m_LambdaOverride == boost::none) {
-        boundingBox.emplace_back(std::log(MIN_REGULARIZER_SCALE * m_TreeImpl->m_Lambda),
-                                 std::log(MAX_REGULARIZER_SCALE * m_TreeImpl->m_Lambda));
+    if (m_TreeImpl->m_RegularizationOverride.lambda() == boost::none) {
+        boundingBox.emplace_back(std::log(m_LambdaSearchInterval(MIN_REGULARIZER_INDEX)),
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeFactory.cc]: No. In fact we arrange for it to be non-zero so the log doesn't explode. It can get so small that it is effectively negligible. For example, if it is much less than 1, for MSE loss, it doesn't really have any effect because we always have at least one row per leaf node.

(I have considered reparameterising so that we control total amount of regularisation and then distribute it out between different terms. This allows lambda to go to zero. However, that change also interacts with the upcoming regulariser based on tree depth. I plan to revisit this down-the-line.)

File: include/core/CLoopProgress.h [hunk 7]
```diff
@@ -46,14 +48,35 @@ class CORE_EXPORT CLoopProgress {
...
-    CLoopProgress(ITR begin, ITR end, const TProgressCallback& recordProgress, double scale = 1.0)
+    CLoopProgress(ITR begin, ITR end, const TProgressCallback& recordProgress = noop, double scale = 1.0)
         : CLoopProgress(std::distance(begin, end), recordProgress, scale) {}
-    CLoopProgress(std::size_t size, const TProgressCallback& recordProgress, double scale = 1.0);
+    CLoopProgress(std::size_t size,
+                  const TProgressCallback& recordProgress = noop,
+                  double scale = 1.0);
+
+    //! Attach a new progress monitor callback.
+    void attach(const TProgressCallback& recordProgress);
```
[the remark this answers is not in the document]
Comment (tveasey) [reply include/core/CLoopProgress.h]: Good suggestion! Done in [79f3b43](https://github.com/elastic/ml-cpp/pull/686/commits/79f3b433cc9085a4e776a08bd8a1ca7f69bb96c9).

File: include/maths/CBoostedTreeFactory.h [hunk 8]
```diff
@@ -121,17 +126,32 @@ class MATHS_EXPORT CBoostedTreeFactory final {
...
+                                                       TScaleRegularization scale) const;
 
     //! Initialize the state for hyperparameter optimisation.
     void initializeHyperparameterOptimisation() const;
 
     //! Get the number of hyperparameter tuning rounds to use.
     std::size_t numberHyperparameterTuningRounds() const;
 
+    //! Setup monitoring for training progress.
+    void setupTrainingProgressMonitoring();
```
[the remark this answers is not in the document]
Comment (tveasey) [reply include/maths/CBoostedTreeFactory.h]: See [aeae71e](https://github.com/elastic/ml-cpp/pull/686/commits/aeae71e0cdd8d13237f71d22d1ba24f221e71fb0).

File: lib/maths/CBoostedTreeFactory.cc [hunk 9]
```diff
@@ -315,6 +280,189 @@ void CBoostedTreeFactory::initializeHyperparameters(core::CDataFrame& frame) con
...
+    // optimisation loop.
+
+    core::CPackedBitVector allTrainingRowsMask{m_TreeImpl->allTrainingRowsMask()};
+
+    double gainPerNode;
+    double totalCurvaturePerNode;
+    std::tie(gainPerNode, totalCurvaturePerNode) =
+        this->estimateTreeGainAndCurvature(frame, allTrainingRowsMask);
+
+    if (gainPerNode > 0.0 && m_TreeImpl->m_RegularizationOverride.gamma() == boost::none) {
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeFactory.cc]: Done in [7111620](https://github.com/elastic/ml-cpp/pull/686/commits/7111620b7fb8a677758c81af6f8a889a1a671a98).

File: lib/maths/CBoostedTreeFactory.cc [hunk 10]
```diff
@@ -315,6 +280,189 @@ void CBoostedTreeFactory::initializeHyperparameters(core::CDataFrame& frame) con
...
+            });
+        m_LambdaSearchInterval = interval.value_or(fallbackInterval) * totalCurvaturePerNode;
+        LOG_TRACE(<< "lambda search interval = ["
+                  << m_LambdaSearchInterval.toDelimited() << "]");
+
+    } else if (m_TreeImpl->m_RegularizationOverride.lambda() == boost::none) {
+        m_TreeImpl->m_RegularizationOverride.lambda(0.0);
+    }
+
+    double scale{
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeFactory.cc]: Done in [1f5d4c0](https://github.com/elastic/ml-cpp/pull/686/commits/1f5d4c0f3135f1a432d2d4b27cfe7953889d35b4)

File: lib/maths/CBoostedTreeFactory.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeFactory.cc]: See [e42208c](https://github.com/elastic/ml-cpp/pull/686/commits/e42208c2eb880fa90fcc29467d1afdef348cf3dc)

File: include/maths/CBoostedTreeImpl.h (hunk 2 above)
Comment (tveasey) [reply include/maths/CBoostedTreeImpl.h]: See [f63c228](https://github.com/elastic/ml-cpp/pull/686/commits/f63c2289883ea053fdf643ec38e19a7885a97c56)

File: include/maths/CBoostedTreeImpl.h (hunk 3 above)
Comment (tveasey) [reply include/maths/CBoostedTreeImpl.h]: [e0dd26e](https://github.com/elastic/ml-cpp/pull/686/commits/e0dd26ede432e16cd19d7ce0961f925db920f2ef)

File: lib/maths/CBoostedTreeFactory.cc [hunk 11]
```diff
@@ -315,6 +280,189 @@ void CBoostedTreeFactory::initializeHyperparameters(core::CDataFrame& frame) con
...
+                                                    static_cast<double>(i)));
+        }
+        double margin{2.0 * std::sqrt(CBasicStatistics::variance(residualMoments)) / curvature};
+        if (logBestRegularizerScale - margin >= leftEndpoint) {
+            interval(MIN_REGULARIZER_INDEX) =
+                std::max(std::pow(0.5, margin), MIN_REGULARIZER_SCALE);
+        }
+        if (logBestRegularizerScale + margin <= rightEndpoint) {
+            interval(MAX_REGULARIZER_INDEX) =
+                std::min(std::pow(2.0, margin), MAX_REGULARIZER_SCALE);
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeFactory.cc]: See [cbe44d6](https://github.com/elastic/ml-cpp/pull/686/commits/cbe44d64570bf1d000cca3e1714df0f29e5e0d8b).

File: lib/maths/CBoostedTreeFactory.cc [hunk 12]
```diff
@@ -315,37 +293,228 @@ void CBoostedTreeFactory::initializeHyperparameters(core::CDataFrame& frame) con
...
+            traverser.haveBadState()) {
             throw std::runtime_error{"failed to restore boosted tree"};
         }
-        frame.resizeColumns(treePtr->m_Impl->m_NumberThreads,
-                            frame.numberColumns() +
-                                treePtr->m_Impl->numberExtraColumnsForTrain());
-        return treePtr;
     } catch (const std::exception& e) {
-        HANDLE_FATAL(<< "Input error: '" << e.what() << "'. Check logs for more details.");
+        throw std::runtime_error{std::string{"Input error: '"} + e.what() + "'"};
```
Comment (tveasey) [inline lib/maths/CBoostedTreeFactory.cc]: Note I changed this to throw. `HANDLE_FATAL` will abort the program, but this is recoverable since we can just start again. The exception is already handled in the runner classes in the api library.

File: lib/maths/CBoostedTreeFactory.cc [hunk 13]
```diff
@@ -300,44 +313,51 @@ void CBoostedTreeFactory::initializeUnsetRegularizationHyperparameters(core::CDa
...
+
+            double initialLambda{totalCurvaturePerNode};
+            auto lambdaStep = [initialLambda](CBoostedTreeImpl& tree, double scale) {
+                tree.m_Regularization.lambda(scale * initialLambda);
+            };
+            m_LambdaSearchInterval = this->lineSearchWithQuadraticApproxToTestError(
+                                             frame, allTrainingRowsMask, lambdaStep)
+                                         .value_or(fallbackInterval) *
+                                     totalCurvaturePerNode;
+            LOG_TRACE(<< "lambda search interval = ["
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeFactory.cc]: The intervals are used when initialising the bounding box for hyperparameter search, i.e. these are written to member variables which are used in `initializeHyperparameterOptimisation`.

File: lib/maths/CBoostedTreeFactory.cc [hunk 14]
```diff
@@ -300,44 +313,51 @@ void CBoostedTreeFactory::initializeUnsetRegularizationHyperparameters(core::CDa
...
+        }
     }
 
-    double scale{
-        static_cast<double>(m_TreeImpl->m_NumberFolds - 1) /
-        static_cast<double>(m_TreeImpl->m_NumberFolds) /
-        ((m_TreeImpl->m_RegularizationOverride.gamma() != boost::none ? 0.0 : 1.0) +
-         (m_TreeImpl->m_RegularizationOverride.lambda() != boost::none ? 0.0 : 1.0))};
+    double freeRegularizationParameters{
+        (m_TreeImpl->m_RegularizationOverride.gamma() != boost::none ? 0.0 : 1.0) +
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeFactory.cc]: Done in [7d2ccb0](https://github.com/elastic/ml-cpp/pull/686/commits/7d2ccb0c4be88a24d40bfdc018188e3d5df25dc2)

File: lib/maths/CBoostedTreeFactory.cc [hunk 15]
```diff
@@ -300,44 +313,51 @@ void CBoostedTreeFactory::initializeUnsetRegularizationHyperparameters(core::CDa
...
+        (m_TreeImpl->m_RegularizationOverride.lambda() != boost::none ? 0.0 : 1.0)};
+    double scale{static_cast<double>(m_TreeImpl->m_NumberFolds - 1) /
+                 static_cast<double>(m_TreeImpl->m_NumberFolds) / freeRegularizationParameters};
 
     if (m_TreeImpl->m_RegularizationOverride.gamma() == boost::none) {
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeFactory.cc]: I added an explanation, for dividing by number of unspecified parameters.

The reason for the `(folds - 1) / folds` is related to the old methodology and is no longer needed. It was a correction to account for the fact that we expected the total loss to be proportional to the count of rows for a random sample.

See [7d2ccb0](https://github.com/elastic/ml-cpp/pull/686/commits/7d2ccb0c4be88a24d40bfdc018188e3d5df25dc2).

**Comment by tveasey** [comment]
I also refactored the line search function along the lines we discussed off line in [937df23](https://github.com/elastic/ml-cpp/pull/686/commits/937df235947e1c0302eb346225aa545b26903d32). I agree this makes the idea clearer: good suggestion! Can you take another look @valeriy42.

File: lib/maths/CBoostedTreeFactory.cc [hunk 16]
```diff
@@ -440,56 +455,59 @@ CBoostedTreeFactory::TOptionalVector CBoostedTreeFactory::lineSearchWithQuadrati
...
-                                                    static_cast<double>(i)));
+                                                    static_cast<double>(i) * stepSize));
         }
         double sigma{std::sqrt(CBasicStatistics::variance(residualMoments))};
-        double logScaleAtThreeSigma{std::sqrt(3.0 * sigma / curvature)};
-        if (logBestRegularizerScale - logScaleAtThreeSigma >= leftEndpoint) {
-            // These are scales > bestRegularizerScale hence 1 / multiplier.
-            interval(MAX_REGULARIZER_INDEX) = std::min(
-                std::pow(1.0 / multiplier, logScaleAtThreeSigma), MAX_REGULARIZER_SCALE);
+        double threeSigmaInterval{std::sqrt(3.0 * sigma / curvature)};
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeFactory.cc]: Right, I have the comment `In particular, we solve curvature * (x - best)^2 = 3 sigma...` which I thought (maybe slightly tangentially) explained this. I feel like this is maybe enough.

File: lib/maths/CBoostedTreeFactory.cc [hunk 17]
```diff
@@ -440,56 +455,59 @@ CBoostedTreeFactory::TOptionalVector CBoostedTreeFactory::lineSearchWithQuadrati
...
     double gradient{params[1]};
     double curvature{params[2]};
     LOG_TRACE(<< "[intercept, slope, curvature] = "
               << core::CContainerPrinter::print(params));
 
-    // Find the scale at the minimum of the least squares quadratic fit
-    // to the test loss in the search interval.
-    double leftEndpoint{0.0};
-    double rightEndpoint{static_cast<double>(INITIAL_REGULARIZER_SEARCH_ITERATIONS - 1)};
+    // Find the scale at the minimum of the least squares quadratic fit to
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeFactory.cc]: [311ade1](https://github.com/elastic/ml-cpp/pull/686/commits/311ade134dbbcf22ad12997c8ad1e75c135aa770)
